File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.java/core/src/main/java/de/unistuttgart/hamster/commands/impl/SetPropertyCommandImpl.java
package de.unistuttgart.hamster.commands.impl;

import de.unistuttgart.hamster.commands.SetPropertyCommand;
import de.unistuttgart.hamster.commands.impl.utils.CommandUtils;

public class SetPropertyCommandImpl extends SetPropertyCommand {

	@Override
	public void execute() {
		try {
			var entity = getEntity();
			var method = CommandUtils.findSetMethod(entity, getPropertyName());
			method.invoke(getEntity(), getNewValue());
		} catch (Exception e) {
			throw new RuntimeException("Error: could not execute SetPropertyCommand '" + getPropertyName() + "' on entity " + getEntity(), e);
		}
	}

	@Override
	public void undo() {
		try {
			var entity = getEntity();
			var method = CommandUtils.findSetMethod(entity, getPropertyName());
			method.invoke(getEntity(), getOldValue());
		} catch (Exception e) {
			throw new RuntimeException("Error: could not execute SetPropertyCommand '" + getPropertyName() + "' on entity " + getEntity(), e);
		}
	}

	@Override
	public void redo() {
		execute();
	}

}

File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/src/impl/SetPropertyCommandImpl.cpp
#include "SetPropertyCommandImpl.h"

#include "Entity.h"

namespace commands {

SetPropertyCommandImpl::SetPropertyCommandImpl() {
}

void SetPropertyCommandImpl::execute() {
    auto entity = getEntity();
    entity->setProperty(getFeatureKey(), getNewValue());
}

void SetPropertyCommandImpl::undo() {
    auto entity = getEntity();
    entity->setProperty(getFeatureKey(), getOldValue());
}

void SetPropertyCommandImpl::redo() {
    execute();
}

void SetPropertyCommandImpl::handleLifeCycleOfValues(deleter deletionFunction) {
    ValueReference oldValue = std::get<ValueReference>(getOldValue());
    ValueReference newValue = std::get<ValueReference>(getNewValue());
    oldValueUniquePtr = std::unique_ptr<void, deleter>(oldValue, deletionFunction);
    newValueUniquePtr = std::unique_ptr<void, deleter>(newValue, deletionFunction);
}

}

